docs/designs/elasticdl_operator.md

```diff
@@ -0,0 +1,211 @@
+# Design for ElasticDL Operator
+
+## Motivation
+
+ElasticDL uses master-worker architecture.
+Each ElasticDL job has a unique master pod.
+The master pod manages the lifecycle of worker pods and controls the training process.
+
+ElasticDL provides a command-line [client tool](https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/designs/client_tool.md)
+to submit a job to a Kubernetes cluster.
+At first, a master pod will be launched.
+Then, the master pod launches worker pods and PS pods if necessary.
+The training process begins once a worker pod becomes ready.
+
+When making ElasticDL as a product of cloud computing,
+we find that we have to address the following two points:
+
+- Job monitoring and management. ElasticDL client tool only launches a job.
+We have to write extra scripts to monitor the pods' status
+and clean pods when a job completes.
+
+- Product compatibility. Current products have deployed some operators of [Kubeflow](https://www.kubeflow.org/),
+such as TF operator and PyTorch operator.
+Many development works have been done when integrating the operators,
+including dashboard, command-line tool, and controllers with rich monitoring functions.
+It's better to reuse the work.
+
+So, we decide to apply the
+[operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/)
+to ElasticDL as well.
+We introduce a CRD to define the workload of ElasticDL jobs.
+Then, we describe each ElastiDL job with a YAML file according to the CRD.
+The custom controller handles the request from the YAML file.
+
+It's a remarkable fact that the controller only launches
+the master pod, and monitors the job.
+The worker pods are still mananged by the master pod.
+The controller does no contribution to fault-torance and elastic scheduling features.
```

Comment:
How does the controller monitor workers launched by the master during training?

Reply:
The controller could use K8S Go API to query the status of worker pods. Besides, it's convenient to introduce monitor tools, like [Prometheus](https://github.com/prometheus/prometheus/), which is also a Go package.